# Lab-Data-Manipulation (PART-1)

In [1]:
import pandas as pd

# Context 

For this lab you'll use a dataset for UFO observations. The objective is to exercise the manipulation of a dataframe, so we'll use the tools we've learned about `reading`, `renaming`, `selecting specific columns`, `filtering based on conditions` and `merging` dataframes to try to better understand our dataset and store an enriched version of our dataset at the end.

Good Luck.

variable	|class|	description
------------|-----|-------------
date_time	|datetime (mdy h:m)	| Date time sighting occurred
city_area	|character	        | City or area of sighting
state	    |character          |	state/region of sighting
country	    |character          |	Country of sighting
ufo_shape	|character          |	UFO Shape
encounter_length	|double     |	Encounter length in seconds
described_encounter_length	|character |	Encounter length as described (eg 1 hour, etc)
description	|character          |	Description of encounter
date_documented	|character      |	Date documented
latitude	|double             |	Latitude
longitude	|double             |	Longitude

## Read the dataset and store it in a dataframe called `ufo`

Pay attention at the file separator.

In [87]:
ufo = pd.read_csv('/Users/alessandra/Documents/Ironhack/Curso/Data/ufo.csv', sep = ';')

In [88]:
ufo

Unnamed: 0        date  year  month  day         date_time  \
0               0  1949-10-10  1949     10   10  10/10/1949 20:30   
1               1  1949-10-10  1949     10   10  10/10/1949 21:00   
2               2  1955-10-10  1955     10   10  10/10/1955 17:00   
3               3  1956-10-10  1956     10   10  10/10/1956 21:00   
4               4  1960-10-10  1960     10   10  10/10/1960 20:00   
...           ...         ...   ...    ...  ...               ...   
80327       80327  2013-09-09  2013      9    9    9/9/2013 21:15   
80328       80328  2013-09-09  2013      9    9    9/9/2013 22:00   
80329       80329  2013-09-09  2013      9    9    9/9/2013 22:00   
80330       80330  2013-09-09  2013      9    9    9/9/2013 22:20   
80331       80331  2013-09-09  2013      9    9    9/9/2013 23:00   

                  city_area state country ufo_shape  encounter_length  \
0                san marcos    tx      us  cylinder            2700.0   
1              lackland afb    tx     NaN     light            7200.0   
2      chester (uk/england)   NaN      gb    circle              20.0   
3                      edna    tx      us    circle              20.0   
4                   kaneohe    hi      us     light             900.0   
...                     ...   ...     ...       ...               ...   
80327             nashville    tn      us     light             600.0   
80328                 boise    id      us    circle            1200.0   
80329                  napa    ca      us     other            1200.0   
80330                vienna    va      us    circle               5.0   
80331                edmond    ok      us     cigar            1020.0   

      described_encounter_length  \
0                     45 minutes   
1                        1-2 hrs   
2                     20 seconds   
3                       1/2 hour   
4                     15 minutes   
...                          ...   
80327                 10 minutes   
80328                 20 minutes   
80329                       hour   
80330                  5 seconds   
80331                 17 minutes   

                                             description date_documented  \
0      This event took place in early fall around 194...       4/27/2004   
1      1949 Lackland AFB&#44 TX.  Lights racing acros...      12/16/2005   
2      Green/Orange circular disc over Chester&#44 En...       1/21/2008   
3      My older brother and twin sister were leaving ...       1/17/2004   
4      AS a Marine 1st Lt. flying an FJ4B fighter/att...       1/22/2004   
...                                                  ...             ...   
80327  Round from the distance/slowly changing colors...       9/30/2013   
80328  Boise&#44 ID&#44 spherical&#44 20 min&#44 10 r...       9/30/2013   
80329                                       Napa UFO&#44       9/30/2013   
80330  Saw a five gold lit cicular craft moving fastl...       9/30/2013   
80331  2 witnesses 2  miles apart&#44 Red &amp; White...       9/30/2013   

        latitude   longitude  
0      29.883056  -97.941111  
1      29.384210  -98.581082  
2      53.200000   -2.916667  
3      28.978333  -96.645833  
4      21.418056 -157.803611  
...          ...         ...  
80327  36.165833  -86.784444  
80328  43.613611 -116.202500  
80329  38.297222 -122.284444  
80330  38.901111  -77.265556  
80331  35.652778  -97.477778  

[80332 rows x 16 columns]

## Check the first 6 columns of the dataframe

In [89]:
ufo.iloc[:,:7]

Unnamed: 0        date  year  month  day         date_time  \
0               0  1949-10-10  1949     10   10  10/10/1949 20:30   
1               1  1949-10-10  1949     10   10  10/10/1949 21:00   
2               2  1955-10-10  1955     10   10  10/10/1955 17:00   
3               3  1956-10-10  1956     10   10  10/10/1956 21:00   
4               4  1960-10-10  1960     10   10  10/10/1960 20:00   
...           ...         ...   ...    ...  ...               ...   
80327       80327  2013-09-09  2013      9    9    9/9/2013 21:15   
80328       80328  2013-09-09  2013      9    9    9/9/2013 22:00   
80329       80329  2013-09-09  2013      9    9    9/9/2013 22:00   
80330       80330  2013-09-09  2013      9    9    9/9/2013 22:20   
80331       80331  2013-09-09  2013      9    9    9/9/2013 23:00   

                  city_area  
0                san marcos  
1              lackland afb  
2      chester (uk/england)  
3                      edna  
4                   kaneohe  
...                     ...  
80327             nashville  
80328                 boise  
80329                  napa  
80330                vienna  
80331                edmond  

[80332 rows x 7 columns]

## Check the shape of your dataframe to see how many rows and columns it has

In [90]:
ufo.shape

(80332, 16)

## Bring the date information to the beginning of the dataframe

If you check the dataframe columns, there are some information of the date at the end of the dataframe. For this task, you should reorder the columns in a way that the first few columns all show the date information. 

*Hint: Use the ufo.columns to see all the column names you have.

In [91]:
ufo.columns

Index(['Unnamed: 0', 'date', 'year', 'month', 'day', 'date_time', 'city_area',
       'state', 'country', 'ufo_shape', 'encounter_length',
       'described_encounter_length', 'description', 'date_documented',
       'latitude', 'longitude'],
      dtype='object')

In [92]:
columns_order = ['date_time', 'date', 'month', 'year', 
                 'city_area', 'state', 'country', 'ufo_shape', 'encounter_length',
                 'described_encounter_length', 'description', 'date_documented','latitude', 'longitude']
ufo[columns_order]

date_time        date  month  year             city_area state  \
0      10/10/1949 20:30  1949-10-10     10  1949            san marcos    tx   
1      10/10/1949 21:00  1949-10-10     10  1949          lackland afb    tx   
2      10/10/1955 17:00  1955-10-10     10  1955  chester (uk/england)   NaN   
3      10/10/1956 21:00  1956-10-10     10  1956                  edna    tx   
4      10/10/1960 20:00  1960-10-10     10  1960               kaneohe    hi   
...                 ...         ...    ...   ...                   ...   ...   
80327    9/9/2013 21:15  2013-09-09      9  2013             nashville    tn   
80328    9/9/2013 22:00  2013-09-09      9  2013                 boise    id   
80329    9/9/2013 22:00  2013-09-09      9  2013                  napa    ca   
80330    9/9/2013 22:20  2013-09-09      9  2013                vienna    va   
80331    9/9/2013 23:00  2013-09-09      9  2013                edmond    ok   

      country ufo_shape  encounter_length described_encounter_length  \
0          us  cylinder            2700.0                 45 minutes   
1         NaN     light            7200.0                    1-2 hrs   
2          gb    circle              20.0                 20 seconds   
3          us    circle              20.0                   1/2 hour   
4          us     light             900.0                 15 minutes   
...       ...       ...               ...                        ...   
80327      us     light             600.0                 10 minutes   
80328      us    circle            1200.0                 20 minutes   
80329      us     other            1200.0                       hour   
80330      us    circle               5.0                  5 seconds   
80331      us     cigar            1020.0                 17 minutes   

                                             description date_documented  \
0      This event took place in early fall around 194...       4/27/2004   
1      1949 Lackland AFB&#44 TX.  Lights racing acros...      12/16/2005   
2      Green/Orange circular disc over Chester&#44 En...       1/21/2008   
3      My older brother and twin sister were leaving ...       1/17/2004   
4      AS a Marine 1st Lt. flying an FJ4B fighter/att...       1/22/2004   
...                                                  ...             ...   
80327  Round from the distance/slowly changing colors...       9/30/2013   
80328  Boise&#44 ID&#44 spherical&#44 20 min&#44 10 r...       9/30/2013   
80329                                       Napa UFO&#44       9/30/2013   
80330  Saw a five gold lit cicular craft moving fastl...       9/30/2013   
80331  2 witnesses 2  miles apart&#44 Red &amp; White...       9/30/2013   

        latitude   longitude  
0      29.883056  -97.941111  
1      29.384210  -98.581082  
2      53.200000   -2.916667  
3      28.978333  -96.645833  
4      21.418056 -157.803611  
...          ...         ...  
80327  36.165833  -86.784444  
80328  43.613611 -116.202500  
80329  38.297222 -122.284444  
80330  38.901111  -77.265556  
80331  35.652778  -97.477778  

[80332 rows x 14 columns]

## Just check if you did it the right way. Take a look at the head of the dataframe again and see if the `ufo` dataframe now is reordered.

In [93]:
ufo.head()

Unnamed: 0        date  year  month  day         date_time  \
0           0  1949-10-10  1949     10   10  10/10/1949 20:30   
1           1  1949-10-10  1949     10   10  10/10/1949 21:00   
2           2  1955-10-10  1955     10   10  10/10/1955 17:00   
3           3  1956-10-10  1956     10   10  10/10/1956 21:00   
4           4  1960-10-10  1960     10   10  10/10/1960 20:00   

              city_area state country ufo_shape  encounter_length  \
0            san marcos    tx      us  cylinder            2700.0   
1          lackland afb    tx     NaN     light            7200.0   
2  chester (uk/england)   NaN      gb    circle              20.0   
3                  edna    tx      us    circle              20.0   
4               kaneohe    hi      us     light             900.0   

  described_encounter_length  \
0                 45 minutes   
1                    1-2 hrs   
2                 20 seconds   
3                   1/2 hour   
4                 15 minutes   

                                         description date_documented  \
0  This event took place in early fall around 194...       4/27/2004   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...      12/16/2005   
2  Green/Orange circular disc over Chester&#44 En...       1/21/2008   
3  My older brother and twin sister were leaving ...       1/17/2004   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...       1/22/2004   

    latitude   longitude  
0  29.883056  -97.941111  
1  29.384210  -98.581082  
2  53.200000   -2.916667  
3  28.978333  -96.645833  
4  21.418056 -157.803611

## Select a piece of your dataframe. Create a new dataframe called `ufo_vars` and select only the following columns of the `ufo` dataframe. 

`year`, `month`, `state`, `country`, `ufo_shape`, `encounter_length`

In [94]:
filtered_columns = ['year', 'month', 'state', 'country', 'ufo_shape', 'encounter_length']
ufo_vars = ufo.loc[:, filtered_columns]

Perform a *.head()* on your result to check if you did it right.

In [95]:
ufo_vars.head()

year  month state country ufo_shape  encounter_length
0  1949     10    tx      us  cylinder            2700.0
1  1949     10    tx     NaN     light            7200.0
2  1955     10   NaN      gb    circle              20.0
3  1956     10    tx      us    circle              20.0
4  1960     10    hi      us     light             900.0

## Rename the variable `encounter_length` to `encounter_seconds`. Keep using the `ufo_vars` dataset for the following tasks, unless specifically specified.

Again, check your results to check if you did it right.

In [96]:
ufo_vars = ufo_vars.rename(columns = {"encounter_length" : "encounter_seconds"})

In [97]:
ufo_vars

year  month state country ufo_shape  encounter_seconds
0      1949     10    tx      us  cylinder             2700.0
1      1949     10    tx     NaN     light             7200.0
2      1955     10   NaN      gb    circle               20.0
3      1956     10    tx      us    circle               20.0
4      1960     10    hi      us     light              900.0
...     ...    ...   ...     ...       ...                ...
80327  2013      9    tn      us     light              600.0
80328  2013      9    id      us    circle             1200.0
80329  2013      9    ca      us     other             1200.0
80330  2013      9    va      us    circle                5.0
80331  2013      9    ok      us     cigar             1020.0

[80332 rows x 6 columns]

## Let's start filtering some records. Create a new dataframe called `ufo_us` and filter the `ufo_vars` dataframe bringing only the results in which the `country` is `"us"`

### Use a mask to perform this task 

A `mask` is nothing more than a condition. This condition is applied to your whole dataframe (or pandas Series).
So for example, if you had a pandas Series with a variable called `Age`, you could create a mask for all people whose `Age` is less than 18 years old using the syntax:

`df['Age'] <= 18`

This would return a pandas series containing `True` and `False` values. For each index, you'd get a value of `True` or `False`.

You could save this mask in a variable, for example:

`condition = (df['Age'] <= 18)`

And then you could use that variable `condition` to select only the cases of the dataframe in which the index returned `True` using:
df[condition].

In [98]:
mask = ufo['country'] == 'us'
ufo_us = ufo_vars.loc[mask, :]

In [99]:
ufo_us

year  month state country ufo_shape  encounter_seconds
0      1949     10    tx      us  cylinder             2700.0
3      1956     10    tx      us    circle               20.0
4      1960     10    hi      us     light              900.0
5      1961     10    tn      us    sphere              300.0
7      1965     10    ct      us      disk             1200.0
...     ...    ...   ...     ...       ...                ...
80327  2013      9    tn      us     light              600.0
80328  2013      9    id      us    circle             1200.0
80329  2013      9    ca      us     other             1200.0
80330  2013      9    va      us    circle                5.0
80331  2013      9    ok      us     cigar             1020.0

[65114 rows x 6 columns]

### Use the .query() method to perform the same task

Remember that the .query() method expects a string. That string should contain the variable of your dataframe without quotation marks and the comparison. For example, if you had a variable called `name`, you'd use a syntax like:
 `df.query('name == "Jack"')`
 
to bring all observations whose column `name` is exactly equal to `"Jack"` (note that Jack should be within quotation marks because a name is a string in this example).

In [100]:
ufo_us = ufo_vars.query('country == "us"')

In [101]:
ufo_us

year  month state country ufo_shape  encounter_seconds
0      1949     10    tx      us  cylinder             2700.0
3      1956     10    tx      us    circle               20.0
4      1960     10    hi      us     light              900.0
5      1961     10    tn      us    sphere              300.0
7      1965     10    ct      us      disk             1200.0
...     ...    ...   ...     ...       ...                ...
80327  2013      9    tn      us     light              600.0
80328  2013      9    id      us    circle             1200.0
80329  2013      9    ca      us     other             1200.0
80330  2013      9    va      us    circle                5.0
80331  2013      9    ok      us     cigar             1020.0

[65114 rows x 6 columns]

See which one do you prefer the most and keep using it for the exercises that follow

## For the `ufo_us` dataframe, select only the cases in which the year is in the first decade (2001-2010). Put that in a variable called `ufo_us_2000`.

Check your results.

In [102]:
ufo_us_2000 = ufo_us.query(' year > 2000 & year < 2011')

In [103]:
ufo_us_2000

year  month state country  ufo_shape  encounter_seconds
102    2001     10    ia      us   triangle              240.0
105    2001     10    ca      us     circle              120.0
106    2001     10    ia      us  rectangle              300.0
107    2001     10    ca      us   changing              900.0
108    2001     10    az      us   triangle               60.0
...     ...    ...   ...     ...        ...                ...
80269  2010      9    mi      us   changing              300.0
80270  2010      9    fl      us       oval             2700.0
80271  2010      9    fl      us       oval             7200.0
80272  2010      9    nm      us   fireball              300.0
80273  2010      9    ca      us     circle              300.0

[31910 rows x 6 columns]

## Try to do the same without the intermediate step of creating the `ufo_us` dataframe. That is, try to filter the dataset for the cases in which the country is "us" and the year is (2001-2010)



*Hint:* You have to make sure all of these conditions are applied simultaneously - using the `and` (or `&`) operator. Try to understand when to use the `and` and the `&` operator.

In [105]:
ufo_vars.query('year > 2000 & year < 2011 & country == "us"')

year  month state country  ufo_shape  encounter_seconds
102    2001     10    ia      us   triangle              240.0
105    2001     10    ca      us     circle              120.0
106    2001     10    ia      us  rectangle              300.0
107    2001     10    ca      us   changing              900.0
108    2001     10    az      us   triangle               60.0
...     ...    ...   ...     ...        ...                ...
80269  2010      9    mi      us   changing              300.0
80270  2010      9    fl      us       oval             2700.0
80271  2010      9    fl      us       oval             7200.0
80272  2010      9    nm      us   fireball              300.0
80273  2010      9    ca      us     circle              300.0

[31910 rows x 6 columns]

## BONUS 1:  Compare the number of triangular UFO occurrances (checking the `ufo_shape` variable) in the US from the year 2001 up to 2010 as compared to the years of 1991 up to 2000.

*Hint: you should expect roughly ~3 times more cases for 2001-2010 than 1991-2000.*

In [106]:
ufo_shape_2000 = ufo_us.query(' year > 2000 & year < 2011 & ufo_shape == "triangle"')
ufo_shape_2000

year  month state country ufo_shape  encounter_seconds
102    2001     10    ia      us  triangle              240.0
108    2001     10    az      us  triangle               60.0
109    2001     10    va      us  triangle               30.0
127    2004     10    nj      us  triangle              180.0
147    2005     10    pa      us  triangle              300.0
...     ...    ...   ...     ...       ...                ...
80195  2004      9    mi      us  triangle              240.0
80231  2008      9    il      us  triangle                7.0
80235  2009      9    co      us  triangle              900.0
80248  2009      9    or      us  triangle              300.0
80256  2009      9    mi      us  triangle              360.0

[3353 rows x 6 columns]

In [108]:
ufo_shape_1990 = ufo_us.query(' year > 1990 & year < 2001 & ufo_shape == "triangle"')
ufo_shape_1990

year  month state country ufo_shape  encounter_seconds
50     1991     10    pa      us  triangle              600.0
64     1996     10    mo      us  triangle                3.0
70     1997     10    ga      us  triangle              300.0
78     1998     10    wa      us  triangle              600.0
291    1994     10    nj      us  triangle              300.0
...     ...    ...   ...     ...       ...                ...
79711  2000      9    az      us  triangle              120.0
79712  2000      9    ny      us  triangle              300.0
79897  1999      9    wa      us  triangle               10.0
79898  1999      9    or      us  triangle               15.0
80139  2000      9    pa      us  triangle               15.0

[1116 rows x 6 columns]

How many rows does each dataset have?

ufo_shape_2000 = **3.353 rows**
/ ufo_shape_1990 = **1.116 rows**

## BONUS 1.1: How many values does each category of `ufo_shape` has. 



In [109]:
ufo_shape_2000.count


<bound method DataFrame.count of        year  month state country ufo_shape  encounter_seconds
102    2001     10    ia      us  triangle              240.0
108    2001     10    az      us  triangle               60.0
109    2001     10    va      us  triangle               30.0
127    2004     10    nj      us  triangle              180.0
147    2005     10    pa      us  triangle              300.0
...     ...    ...   ...     ...       ...                ...
80195  2004      9    mi      us  triangle              240.0
80231  2008      9    il      us  triangle                7.0
80235  2009      9    co      us  triangle              900.0
80248  2009      9    or      us  triangle              300.0
80256  2009      9    mi      us  triangle              360.0

[3353 rows x 6 columns]>

In [110]:
ufo_shape_1990.count


<bound method DataFrame.count of        year  month state country ufo_shape  encounter_seconds
50     1991     10    pa      us  triangle              600.0
64     1996     10    mo      us  triangle                3.0
70     1997     10    ga      us  triangle              300.0
78     1998     10    wa      us  triangle              600.0
291    1994     10    nj      us  triangle              300.0
...     ...    ...   ...     ...       ...                ...
79711  2000      9    az      us  triangle              120.0
79712  2000      9    ny      us  triangle              300.0
79897  1999      9    wa      us  triangle               10.0
79898  1999      9    or      us  triangle               15.0
80139  2000      9    pa      us  triangle               15.0

[1116 rows x 6 columns]>

# Lab-Data-Manipulation (PART-2)

The second part of this lab consists of grouping and merging results.

# Grouping up the results. 

## Let's calculate the average of the encounter for each country.

We should now group the results by the country column to see what is the mean encounter_seconds for each country. Do this using the `groupby` method of your dataframe `ufo_vars`. What is the average of the encounter for the us? And for Canada?

Remember that after grouping by a column, you have to specify a `aggregating function`. If you don't do that, the results of the groupby will only be a `groupby` pandas object. For this case, we want the aggregating function to be the `mean` function and then the results will appear for us.

Also remember that **if you don't** specify the `as_index=False` argument, the variables you use to group are going to become your new indexes.

In [111]:
import pandas as pd

In [112]:
ufo_vars.groupby(by = ['country'], as_index = False).agg('mean')

country         year     month  encounter_seconds
0      au  2002.693309  6.128253        3806.469238
1      ca  2003.694333  6.861667       28859.437007
2      de  2001.542857  6.247619       24255.980952
3      gb  2003.648294  6.562205       66061.321207
4      us  2004.090057  6.861597        5800.192049

mean US = **5800.192049** / mean CA = **28859.437007**

## Perform the same task, but instead of calculating the mean, count how many occurrances for each country.

For this case, the aggregating function should be the `count` function. Try to understand the results for each column.

In [113]:
ufo_vars.groupby(by = ['country'], as_index = False).agg('count')

country   year  month  state  ufo_shape  encounter_seconds
0      au    538    538     10        527                538
1      ca   3000   3000   2985       2955               3000
2      de    105    105      0        103                105
3      gb   1905   1905     12       1855               1905
4      us  65114  65114  65114      63561              65112

## Perform the same task, but instead of calculating the mean, use the `.describe()` aggregating function to see the effects.

The describe aggregating function will show you several important statistics for the grouped results, such as `mean`, `median`, `standard deviation`, `count`, `max`, `min`, and so on.

*Hint: If it starts to get difficult to see the results, you can tranpose the resulting dataframe by just putting a `.T` at the end.*

In [114]:
ufo_vars.groupby(by=['country']).describe()

year                                                          \
           count         mean        std     min     25%     50%     75%   
country                                                                    
au         538.0  2002.693309   8.501768  1958.0  2001.0  2004.0  2007.0   
ca        3000.0  2003.694333   9.382976  1936.0  2002.0  2005.0  2010.0   
de         105.0  2001.542857  12.411245  1962.0  2000.0  2006.0  2009.0   
gb        1905.0  2003.648294   8.406754  1943.0  2002.0  2005.0  2009.0   
us       65114.0  2004.090057  10.385694  1910.0  2001.0  2007.0  2011.0   

                   month            ...             encounter_seconds  \
            max    count      mean  ...   75%   max             count   
country                             ...                                 
au       2014.0    538.0  6.128253  ...   9.0  12.0             538.0   
ca       2014.0   3000.0  6.861667  ...   9.0  12.0            3000.0   
de       2014.0    105.0  6.247619  ...   8.0  12.0             105.0   
gb       2014.0   1905.0  6.562205  ...   9.0  12.0            1905.0   
us       2014.0  65114.0  6.861597  ...  10.0  12.0           65112.0   

                                                                           
                 mean           std   min   25%    50%    75%         max  
country                                                                    
au        3806.469238  5.434101e+04  0.50  30.0  180.0  600.0   1209600.0  
ca       28859.437007  1.511732e+06  0.02  20.0  120.0  600.0  82800000.0  
de       24255.980952  1.859988e+05  2.00  30.0  180.0  600.0   1814400.0  
gb       66061.321207  2.314714e+06  0.30  30.0  180.0  600.0  97836000.0  
us        5800.192049  4.045197e+05  0.01  30.0  180.0  600.0  66276000.0  

[5 rows x 24 columns]

## Now, let's get deeper in the analysis and group the results not only by country. But by `country` and `year`

### Check the values of the mean and count for the `encounter_seconds` variable for each year. Can you see some discrepancy?

*Hint*: If you want, you can use the `ufo_us` dataset just to see the results for the united states. You could also (in a hacky way) perform the filter right before the groupby operation if you wanted.

In [116]:
ufo_vars.groupby(by=['country', 'year']).agg(mean_encounter_seconds = ('encounter_seconds', 'mean'),
                                             count_encounter_seconds = ('encounter_seconds' , 'count'))

mean_encounter_seconds  count_encounter_seconds
country year                                                 
au      1958             2700.000000                        1
        1960              180.000000                        1
        1967              300.000000                        1
        1968              300.000000                        1
        1972              403.333333                        3
...                              ...                      ...
us      2010             2271.987232                     3548
        2011             2544.292555                     4379
        2012            10640.644916                     6320
        2013             1266.387888                     6056
        2014              782.372979                     1964

[282 rows x 2 columns]

# BONUS 2: Which months are the ones with the highest numbers of occurrences?



In [117]:
ufo_vars.groupby(by=['month'])['year'].agg(['count'])

count
month       
1       5689
2       4665
3       5450
4       5527
5       5293
6       8152
7       9520
8       8636
9       7589
10      7407
11      6739
12      5665

**July, August and June are the months with the highest numbers of occurrences**

# BONUS 3: Finally, you gathered information about the UFO dataset. Using your last result, try to bring that information for your original dataset.

1. Store the results of your previous analysis (the mean value for the encounter_seconds for each year and each country) in a dataframe called `avg_results`. 

In [118]:
avg_results = ufo_vars.groupby(by=['country', 'year']).agg(encounter_seconds = ('encounter_seconds', 'mean'))

In [119]:
avg_results

encounter_seconds
country year                   
au      1958        2700.000000
        1960         180.000000
        1967         300.000000
        1968         300.000000
        1972         403.333333
...                         ...
us      2010        2271.987232
        2011        2544.292555
        2012       10640.644916
        2013        1266.387888
        2014         782.372979

[282 rows x 1 columns]

2. Rename the column named `encounter_seconds` to `avg_encounter_seconds`.

In [120]:
avg_results.rename(columns={'encounter_seconds' : 'avg_encounter_seconds'})

avg_encounter_seconds
country year                       
au      1958            2700.000000
        1960             180.000000
        1967             300.000000
        1968             300.000000
        1972             403.333333
...                             ...
us      2010            2271.987232
        2011            2544.292555
        2012           10640.644916
        2013            1266.387888
        2014             782.372979

[282 rows x 1 columns]

3. Use the pd.merge( ... ) function to bring that new collected information to your original dataset.
The pd.merge() function requires several arguments, let's understand the most important ones.

`left` is the dataframe you want to bring information **to** - the table on the left. In this case, this will be our original dataframe called `ufo`

`right` is the dataframe you want to bring information **from** - the table on the right. In this case, this will be our resulting dataframe `avg_results`.

`on` is the key you want to perform the merge. That is, if those values are **exactly equal** in both dataframes, then the information will be brought.

Put your results on a dataframe called `merged_ufo`

In [121]:
merged_ufo = pd.merge(left = ufo, right = avg_results, on = ['country', 'year'])
merged_ufo

Unnamed: 0        date  year  month  day         date_time  \
0               0  1949-10-10  1949     10   10  10/10/1949 20:30   
1            1706  1949-10-15  1949     10   15  10/15/1949 20:00   
2           35615  1949-04-10  1949      4   10   4/10/1949 15:00   
3           41489  1949-05-01  1949      5    1    5/1/1949 14:00   
4           42287  1949-05-15  1949      5   15   5/15/1949 06:30   
...           ...         ...   ...    ...  ...               ...   
70657       65598  1943-08-13  1943      8   13   8/13/1943 22:00   
70658       71056  1973-08-03  1973      8    3    8/3/1973 22:30   
70659       72796  1999-09-10  1999      9   10   9/10/1999 21:10   
70660       74514  1934-09-15  1934      9   15   9/15/1934 15:30   
70661       74515  1936-09-15  1936      9   15   9/15/1936 19:00   

                             city_area state country  ufo_shape  \
0                           san marcos    tx      us   cylinder   
1                            baltimore    md      us      cigar   
2                              seattle    wa      us       disk   
3                    oak ridge (drive)    wv      us      other   
4                                grays    sc      us      light   
...                                ...   ...     ...        ...   
70657          portsmouth (uk/england)   NaN      gb     sphere   
70658           schwetzingen (germany)   NaN      de    unknown   
70659    weiden (ne bavaria) (germany)   NaN      de      light   
70660  valley city (6.5 miles east of)    nd      us       disk   
70661                  albany (canada)    pe      ca  rectangle   

       encounter_length described_encounter_length  \
0                2700.0                 45 minutes   
1                 180.0                2-3 minutes   
2                 600.0                ten minutes   
3                  40.0                  40seconds   
4                 300.0                   2-5 min.   
...                 ...                        ...   
70657             600.0                     10mins   
70658              60.0                 60 seconds   
70659               6.0                5-6 seconds   
70660               5.0                  5 seconds   
70661            1200.0                 2o minutes   

                                             description date_documented  \
0      This event took place in early fall around 194...       4/27/2004   
1      It was around 8:PM and I was walking home&#44i...        3/4/2008   
2                            UFO Hovers over Seattle sky       4/27/2004   
3      log shape whooshing object observed near charl...       3/19/2002   
4      I am not sure of the exact date or time of thi...      10/31/2003   
...                                                  ...             ...   
70657  vertical lines of 5-6 crescent moons&#44 some ...       4/16/2005   
70658  A light traveling at a high rate of speed hori...        8/5/2012   
70659  2 bright light &quot;events&quot; happening ab...       9/12/1999   
70660  Two brothers observe&#44 hear&#44 three black ...       6/12/2008   
70661  I WAS PLAYING AT MY FRIEND&#39S HOME AFTER SUP...       3/19/2009   

        latitude   longitude  encounter_seconds  
0      29.883056  -97.941111         523.333333  
1      39.290278  -76.612500         523.333333  
2      47.606389 -122.330833         523.333333  
3      38.075000  -81.109722         523.333333  
4      32.673611  -81.020556         523.333333  
...          ...         ...                ...  
70657  50.796700   -1.083333         600.000000  
70658  49.378056    8.581944          60.000000  
70659  51.066667   11.366667           6.000000  
70660  46.923333  -98.002778           5.000000  
70661  52.250000  -81.616667        1200.000000  

[70662 rows x 17 columns]

Check how many rows the final result has and try to explain it. Did the dataset get smaller? Bigger? Or the same? Can you explain why? 

*hint: If you've found something weird, try looking back at your original dataframe and find something interesting on it.*

In [122]:
print(ufo.shape, merged_ufo.shape)
merged_ufo_analyze = pd.merge(left = ufo, right = avg_results, on = ['country', 'year'], how='outer')
print(merged_ufo_analyze.shape)

(80332, 16) (70662, 17)
(80332, 17)


In [123]:
ufo.loc[:, 'country'].isna().sum()

9670

## Store the results into a new csv file called `ufo_enriched.csv`. 

Don't forget to use `index=False`.

In [124]:
merged_ufo.to_csv('/Users/alessandra/Documents/Ironhack/Curso/Data/ufo_enriched_tab.csv',index=False)